### PR TITLE
[JENKINS-41904] NPE in form validation when configuring an external library

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-41904](https://issues.jenkins-ci.org/browse/JENKINS-41904)

Corrects a regression apparently introduced in #94; `checkScanCredentials` actually _was_ checking for a null `context` to some extent, but not enough.

There is generally no `SCMSourceOwner` for an `SCMSource` being used from a library. Impossible for a global library. For a library scoped to a folder, there would be an `Item` always (may or may not be an `SCMSourceOwner`), and actually `Item` is all we care about.

@reviewbybees

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -57,7 +57,7 @@
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>credentials</artifactId>
-            <version>2.1.8</version>
+            <version>2.1.12</version>
```
Comment (jglick) [inline pom.xml]: Picks up https://github.com/jenkinsci/credentials-plugin/pull/68.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 2]
```diff
@@ -78,7 +75,12 @@ private Connector() {
         throw new IllegalAccessError("Utility class");
     }
 
-    public static ListBoxModel listScanCredentials(SCMSourceOwner context, String apiUri) {
+    /** Binary compatibility. */
+    public static ListBoxModel listScanCredentials(@CheckForNull SCMSourceOwner context, String apiUri) {
+        return listScanCredentials((Item) context, apiUri);
+    }
+
+    public static ListBoxModel listScanCredentials(@CheckForNull Item context, String apiUri) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: Adding nullability annotations, though I could not get FindBugs to report errors anyway…not sure what kind of analysis it is doing.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 3]
```diff
@@ -109,7 +116,8 @@ public static FormValidation checkScanCredentials(SCMSourceOwner context, String
             if (!found) {
                 return FormValidation.error("Credentials not found");
             }
-            if (!(context.hasPermission(Item.CONFIGURE)
+            if (context != null && !(
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: The main fix. Note that we already did a null/permission check at the start of this method.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java [hunk 4]
```diff
@@ -234,17 +255,6 @@ private static Proxy getProxy(@Nonnull String host) {
     }
 
     /**
-     * @param config url and creds id to be hashed
-     *
-     * @return unique id for folder name to create cache inside of base cache dir
-     */
-    private static String hashed(GitHubServerConfig config) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java]: Dead code.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java [hunk 5]
```diff
@@ -593,17 +592,17 @@ public SCMNavigator newInstance(String name) {
         }
 
         @Restricted(NoExternalUse.class)
-        public FormValidation doCheckScanCredentialsId(@AncestorInPath SCMSourceOwner context,
+        public FormValidation doCheckScanCredentialsId(@CheckForNull @AncestorInPath Item context,
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java]: Patching this file just for completeness.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java [hunk 6]
```diff
@@ -1225,7 +1223,7 @@ public FormValidation doCheckIncludes(@QueryParameter String value) {
         }
 
         @Restricted(NoExternalUse.class)
-        public FormValidation doCheckScanCredentialsId(@AncestorInPath SCMSourceOwner context,
+        public FormValidation doCheckScanCredentialsId(@CheckForNull @AncestorInPath Item context,
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: Switching to `Item` should ensure that form validation & completion will work even on a library scoped to a plain folder when the credentials are also folder-scoped.

File: pom.xml (hunk 1 above)
Comment (andresrc) [inline pom.xml]: The change is available in 2.1.9, 2.1.12 forces picking a version less than a week old.

Comment (jglick) [inline pom.xml]: As a rule when updating a dep I would just go to the newest version unless there is a particular reason to hold back, but if you prefer 2.1.9 I could downgrade.

Comment (andresrc) [inline pom.xml]: No need to downgrade, just a comment, thx!

**Comment by stephenc** [comment]
I will take care of merging this after #124 as I want to ensure that we do not have additional API usage creeping in without the required rate limit checking

**Pull request merged**